src/lib.rs

```diff
@@ -457,34 +520,224 @@ impl Cascade {
+            exclude_set.set.push(generator);
+        }
+    }
+
+    /// `collect_exclude_set` merges an `ExcludeSet` into the internal storage of the CascadeBuilder.
+    pub fn collect_exclude_set(&mut self, exclude_set: &mut ExcludeSet) {
+        match self.status {
+            BuildStatus(0, ref mut cap) if *cap >= exclude_set.size => *cap -= exclude_set.size,
+            _ => panic!("capacity violation"),
+        }
+        self.to_exclude.append(&mut exclude_set.set);
+    }
+
+    fn push_layer(&mut self) -> Result<(), CascadeError> {
+        // At even layers we encode elements of to_include. At odd layers we encode elements of
+        // to_exclude. In both cases, we track false positives by filtering the complement of the
+        // encoded set through the newly produced bloom filter.
+        let at_even_layer = self.filters.len() % 2 == 0;
+        let (to_encode, to_filter) = match at_even_layer {
+            true => (&mut self.to_include, &mut self.to_exclude),
+            false => (&mut self.to_exclude, &mut self.to_include),
+        };
+
+        let mut bloom = Bloom::new_crlite_bloom(to_encode.len(), to_filter.len(), false);
+
+        // temporarily take self.salt since we need an immutable reference for bloom.insert
+        let salt = std::mem::take(&mut self.salt);
+
+        to_encode.iter_mut().for_each(|x| {
+            x.next_layer();
+            bloom.insert(x, &salt)
+        });
+
+        let mut delta = to_filter.len();
+        to_filter.retain_mut(|x| {
+            x.next_layer();
+            bloom.has(x, &salt)
+        });
+        delta -= to_filter.len();
+
+        if delta == 0 {
+            // Check for collisions between the |to_encode| and |to_filter| sets.
+            // The implementation of PartialEq for CascadeIndexGenerator will successfully
+            // identify cases where the user called |include(item)| and |exclude(item)| for the
+            // same item. It will not identify collisions in the underlying hash function.
+            for x in to_encode.iter_mut() {
+                if to_filter.contains(x) {
+                    return Err(CascadeError::Collision);
+                }
+            }
+        }
+
+        self.salt = salt;
+        self.filters.push(bloom);
+        Ok(())
+    }
+
+    pub fn finalize(mut self) -> Result<Box<Cascade>, CascadeError> {
+        match self.status {
+            BuildStatus(0, 0) => (),
+            _ => panic!("capacity violation"),
+        }
+
+        loop {
+            if self.to_exclude.is_empty() {
+                break;
+            }
+            self.push_layer()?;
+
+            if self.to_include.is_empty() {
+                break;
+            }
+            self.push_layer()?;
+        }
+
+        Ok(Box::new(Cascade {
+            filters: self.filters,
+            salt: self.salt,
+            hash_algorithm: self.hash_algorithm,
+            inverted: false,
```

Comment:
If the add-ons team wants to use this code, it'll need to support inverted filters, but that should be easy to add later.

Reply:
The easiest way to support inverted filters is to have the caller swap "included" and "excluded" and require them to flip the "inverted" bit on the resulting cascade. I added `Cascade::invert` to support this.